python/ray/serve/schema.py

```diff
@@ -1042,3 +1042,22 @@ def _get_status(self) -> ServeStatus:
                 for app_name, app in self.applications.items()
             },
         )
+
+    def dict(self, *args, **kwargs) -> Dict[str, Any]:
+        """Generates dictionary with user facing data."""
+        values = super().dict(*args, **kwargs)
+
+        # `serialized_policy_def` is only used internally and should not be exposed to
+        # the REST api. This method iteratively removes it from each autoscaling config
+        # if exists.
```

Comment:
what'll happen here if the user passed a class definition directly for the autoscaling policy?

Reply:
`policy` should carry the callable, `serialized_policy_def` should continue to be hidden away from users

Reply:
Right, but a callable is not JSON-serializable, so won't this error?

Reply:
No error, added test cases for none, string, and callable policies https://github.com/ray-project/ray/pull/42339/files#diff-1eaedadea6f2e3382aee6726429ebaad94801e3c55592b3e9f84a0d1b7cc1a14R80

Reply:
What will the JSON serialized output be if the user passes in a callable?

Reply:
Oh the policy is actually not part of the response.
```{'controller_info': {'node_id': '055b82cdbc268a384bcc12d2a10aeff0e7177d8cca333dd0eb5b83c5', 'node_ip': '127.0.0.1', 'actor_id': '4d30a0a6062580f1c0b4986e01000000', 'actor_name': 'SERVE_CONTROLLER_ACTOR', 'worker_id': '08c197b9680a2f2c80ce7b947ae14112c5b8d9e9c61f3f5559967734', 'log_file_path': '/serve/controller_39051.log'}, 'proxy_location': <ProxyLocation.HeadOnly: 'HeadOnly'>, 'http_options': {'host': '0.0.0.0'}, 'grpc_options': {}, 'proxies': {'055b82cdbc268a384bcc12d2a10aeff0e7177d8cca333dd0eb5b83c5': {'node_id': '055b82cdbc268a384bcc12d2a10aeff0e7177d8cca333dd0eb5b83c5', 'node_ip': '127.0.0.1', 'actor_id': '6e726f772bab372cca14642e01000000', 'actor_name': 'SERVE_PROXY_ACTOR-055b82cdbc268a384bcc12d2a10aeff0e7177d8cca333dd0eb5b83c5', 'worker_id': 'b20d9e2822e8c11375d4d7c0cbf7675666cb6af20c76fc6b6a1f3e19', 'log_file_path': '/serve/proxy_127.0.0.1.log', 'status': <ProxyStatus.HEALTHY: 'HEALTHY'>}}, 'applications': {'default': {'name': 'default', 'route_prefix': '/', 'docs_path': None, 'status': <ApplicationStatus.RUNNING: 'RUNNING'>, 'message': '', 'last_deployed_time_s': 1705093256.681717, 'deployed_app_config': None, 'deployments': {'autoscaling_app': {'name': 'autoscaling_app', 'status': <DeploymentStatus.HEALTHY: 'HEALTHY'>, 'status_trigger': <DeploymentStatusTrigger.CONFIG_UPDATE_COMPLETED: 'CONFIG_UPDATE_COMPLETED'>, 'message': '', 'deployment_config': {'name': 'autoscaling_app', 'max_concurrent_queries': 100, 'user_config': None, 'autoscaling_config': {'min_replicas': 1, 'max_replicas': 10, 'target_num_ongoing_requests_per_replica': 1.0, 'metrics_interval_s': 10.0, 'look_back_period_s': 30.0, 'smoothing_factor': 1.0, 'upscale_smoothing_factor': None, 'downscale_smoothing_factor': None, 'downscale_delay_s': 600.0, 'upscale_delay_s': 30.0}, 'graceful_shutdown_wait_loop_s': 2.0, 'graceful_shutdown_timeout_s': 20.0, 'health_check_period_s': 10.0, 'health_check_timeout_s': 30.0, 'ray_actor_options': {'runtime_env': {}, 'num_cpus': 1.0}}, 'replicas': [{'node_id': '055b82cdbc268a384bcc12d2a10aeff0e7177d8cca333dd0eb5b83c5', 'node_ip': '127.0.0.1', 'actor_id': '52d96269eaae2b143affb19f01000000', 'actor_name': 'SERVE_REPLICA::default#autoscaling_app#zuqs6gas', 'worker_id': '3e16bb2e8daf6af1faf5b0cccd7857bc0b76acabacb5199175dfc547', 'log_file_path': '/serve/replica_default_autoscaling_app_zuqs6gas.log', 'replica_id': 'default#autoscaling_app#zuqs6gas', 'state': <ReplicaState.RUNNING: 'RUNNING'>, 'pid': 39072, 'start_time_s': 1705093256.7909498}]}}}}, 'target_capacity': None}
```
let me see why that is

Reply:
ok this is due to we do not include `policy` in the `AutoscalingConfig` https://github.com/ray-project/ray/blob/f76081f0f4d07eb0bf771aac8e7ceb8069a60dec/src/ray/protobuf/serve.proto#L65 so when it's serialized the data back to deployment info, it doesn't contain that field. Kinda try out a few things, and just decided to add a new user facing field `policy_name` in the status. Let me know what you think. This is how it looks after
```
{ 'applications': { 'default': { 'deployed_app_config': None,
                                 'deployments': { 'autoscaling_app': { 'deployment_config': { 'autoscaling_config': { 'downscale_delay_s': 600.0,
                                                                                                                      'downscale_smoothing_factor': None,
                                                                                                                      'look_back_period_s': 30.0,
                                                                                                                      'max_replicas': 10,
                                                                                                                      'metrics_interval_s': 10.0,
                                                                                                                      'min_replicas': 1,
                                                                                                                      'policy_name': 'replica_queue_length_autoscaling_policy',  ## <---new field
                                                                                                                      'smoothing_factor': 1.0,
                                                                                                                      'target_num_ongoing_requests_per_replica': 1.0,
                                                                                                                      'upscale_delay_s': 30.0,
                                                                                                                      'upscale_smoothing_factor': None},
                                                                                              'graceful_shutdown_timeout_s': 20.0,
                                                                                              'graceful_shutdown_wait_loop_s': 2.0,
                                                                                              'health_check_period_s': 10.0,
                                                                                              'health_check_timeout_s': 30.0,
                                                                                              'max_concurrent_queries': 100,
                                                                                              'name': 'autoscaling_app',
                                                                                              'ray_actor_options': { 'num_cpus': 1.0,
                                                                                                                     'runtime_env': { }},
                                                                                              'user_config': None},
                                                                       'message': '',
                                                                       'name': 'autoscaling_app',
                                                                       'replicas': [ { 'actor_id': '2777ad457fa57bbb2480273301000000',
                                                                                       'actor_name': 'SERVE_REPLICA::default#autoscaling_app#qk092qkb',
                                                                                       'log_file_path': '/serve/replica_default_autoscaling_app_qk092qkb.log',
                                                                                       'node_id': '2f61e8786194040ab4a43972a15dc2d1de41e368a36a1061bbbc9e38',
                                                                                       'node_ip': '127.0.0.1',
                                                                                       'pid': 43655,
                                                                                       'replica_id': 'default#autoscaling_app#qk092qkb',
                                                                                       'start_time_s': 1705099874.064668,
                                                                                       'state': <ReplicaState.RUNNING: 'RUNNING'>,
                                                                                       'worker_id': 'a91e56f6fcd2d97bf8ec0a1cb0d036a3066bd4527ca5ccd9ba9a0821'}],
                                                                       'status': <DeploymentStatus.HEALTHY: 'HEALTHY'>,
                                                                       'status_trigger': <DeploymentStatusTrigger.CONFIG_UPDATE_COMPLETED: 'CONFIG_UPDATE_COMPLETED'>}},
                                 'docs_path': None,
                                 'last_deployed_time_s': 1705099873.9227,
                                 'message': '',
                                 'name': 'default',
                                 'route_prefix': '/',
                                 'status': <ApplicationStatus.RUNNING: 'RUNNING'>}},
  'controller_info': { 'actor_id': 'abe00c23c7d83f4b75ad022a01000000',
                       'actor_name': 'SERVE_CONTROLLER_ACTOR',
                       'log_file_path': '/serve/controller_43653.log',
                       'node_id': '2f61e8786194040ab4a43972a15dc2d1de41e368a36a1061bbbc9e38',
                       'node_ip': '127.0.0.1',
                       'worker_id': '0531e60d5965a616fca6baee32ac311bdfdabf420a744b77f8b6ee28'},
  'grpc_options': {},
  'http_options': {'host': '0.0.0.0'},
  'proxies': { '2f61e8786194040ab4a43972a15dc2d1de41e368a36a1061bbbc9e38': { 'actor_id': 'e3a47ef12f2a2c8848e5875e01000000',
                                                                             'actor_name': 'SERVE_PROXY_ACTOR-2f61e8786194040ab4a43972a15dc2d1de41e368a36a1061bbbc9e38',
                                                                             'log_file_path': '/serve/proxy_127.0.0.1.log',
                                                                             'node_id': '2f61e8786194040ab4a43972a15dc2d1de41e368a36a1061bbbc9e38',
                                                                             'node_ip': '127.0.0.1',
                                                                             'status': <ProxyStatus.HEALTHY: 'HEALTHY'>,
                                                                             'worker_id': 'dd0916116d09ae0bb83a8b7b53459465bb60101a04048805a14105ca'}},
  'proxy_location': <ProxyLocation.HeadOnly: 'HeadOnly'>,
  'target_capacity': None}
```

Reply:
Including a field in the response that isn't part of the input config schema is a little odd and breaks the existing behavior where this output can be directly used as input to apply the config

I'd propose the following:

- Include the `autoscaling_policy` field.
- If the user gave an import path, we give it back.
- If the user gave a `Callable`, we get its import path to the best of our ability (`f"{cls.__module__}.{cls.__name__}") and return it